crates/sema/src/ast_passes/utils.rs

```diff
@@ -0,0 +1,70 @@
+use solar_ast::{visit::Visit, Stmt};
+use solar_data_structures::Never;
+use solar_interface::diagnostics::DiagCtxt;
+use std::ops::ControlFlow;
+
+struct VariableDeclarationAsLoopBodyChecker<'ast, 'sess> {
+    visited_block: bool,
+    #[allow(dead_code)]
+    body: &'ast &'ast mut Stmt<'ast>,
+    dcx: &'sess DiagCtxt,
+}
+
+impl<'ast, 'sess> VariableDeclarationAsLoopBodyChecker<'ast, 'sess> {
+    pub(crate) fn new(body: &'ast &'ast mut Stmt<'ast>, dcx: &'sess DiagCtxt) -> Self {
+        Self { visited_block: false, body, dcx }
+    }
+
+    /// Returns the diagnostics context.
+    #[inline]
+    fn dcx(&self) -> &'sess DiagCtxt {
+        self.dcx
+    }
+}
+
+impl<'ast> Visit<'ast> for VariableDeclarationAsLoopBodyChecker<'ast, '_> {
+    type BreakValue = Never;
+
+    fn visit_block(
+        &mut self,
+        block: &'ast solar_ast::Block<'ast>,
+    ) -> ControlFlow<Self::BreakValue> {
+        self.visited_block = true;
+        self.walk_block(block)
+    }
+
+    fn visit_variable_definition(
+        &mut self,
+        var: &'ast solar_ast::VariableDefinition<'ast>,
+    ) -> ControlFlow<Self::BreakValue> {
+        if !self.visited_block {
+            self.dcx()
+                .err("variable declarations are not allowed as the body of a loop")
+                .span(var.span)
+                .help("wrap the statement in a block (`{ ... }`)")
+                .emit();
+        }
+        self.walk_variable_definition(var)
+    }
+
+    fn visit_stmt(&mut self, stmt: &'ast solar_ast::Stmt<'ast>) -> ControlFlow<Self::BreakValue> {
+        match &stmt.kind {
+            solar_ast::StmtKind::While(..)
+            | solar_ast::StmtKind::DoWhile(..)
+            | solar_ast::StmtKind::For { .. } => {
+                return ControlFlow::Continue(());
+            }
+            _ => {}
+        };
+        self.walk_stmt(stmt)
+    }
+}
+
+pub(crate) fn check_if_loop_body_is_a_variable_declaration<'ast>(
```

Comment:
This is much easier than you think, you can simply make this a method on `AstValidator` that checks that a statement is one of the variable declaration statements.

See the equivalent code in solc https://github.com/ethereum/solidity/blob/a62a455039120da834fd5faac2e0db8fc222c10d/libsolidity/analysis/SyntaxChecker.cpp#L190-L222
`if (dynamic_cast<VariableDeclarationStatement const*>)` is equivalent to `matches!(stmt, ast::StmtKind::DeclSingle(..) | ast::StmtKind::DeclMulti(..))`; you can also add this check as a method on `StmtKind`.